subxt/src/backend/unstable/mod.rs

```diff
@@ -145,34 +148,51 @@ impl<T: Config> UnstableBackend<T> {
         I: IntoIterator<Item = follow_stream_unpin::BlockRef<T::Hash>> + Send + 'static,
         <I as IntoIterator>::IntoIter: Send,
     {
-        let sub_id = get_subscription_id(&self.follow_handle).await?;
-        let sub_id = Arc::new(sub_id);
         let methods = self.methods.clone();
-        let headers = self.follow_handle.subscribe().events().flat_map(move |ev| {
-            let sub_id = sub_id.clone();
-            let methods = methods.clone();
-
-            let block_refs = f(ev).into_iter();
+        let follow_handle = self.follow_handle.clone();
+        let sub_id = get_subscription_id(&follow_handle).await?;
 
-            futures::stream::iter(block_refs).filter_map(move |block_ref| {
-                let sub_id = sub_id.clone();
+        // TODO(niklasad1): get rid of the Arc<RwLock> if possible...
+        let headers = self
+            .follow_handle
+            .subscribe()
+            .scan(Arc::new(RwLock::new(sub_id)), |id, v| {
+                let sub_id = id.clone();
+                async { Some((sub_id, v)) }
+            })
+            .filter_map(|(sub_id, v)| async {
+                match v {
+                    FollowStreamMsg::Event(ev) => Some((sub_id, ev)),
+                    FollowStreamMsg::Ready(id) => {
+                        *sub_id.write().unwrap() = id;
+                        None
+                    }
+                }
+            })
```

Comment:
If we wanted to avoid modifying the underlying stuff and get rid of the lock, maybe something like this would work:

```rust
self.follow_handle.subscribe().scan(None, |sub_id, ev| {
    if let FollowStreamMsg::Ready(id) = &ev {
        *sub_id = Some(id);
        Some((sub_id.expect("sub_id should always be present"), ev))
    } else {
        Some((sub_id.expect("sub_id should always be present"), ev))
    }
}).filter_map(|(sub_id, ev)| {
    if let FollowStreamMsg::Event(ev) = ev {
        Some((sub_id, ev))
    } else {
        None
    }
})
```

`follow_handle.subscribe()` should always return a `Ready` event with a sub_id in it before anything else.

Doing the above also means you can ditch the `let sub_id = get_subscription_id(&follow_handle).await?;` line, which is basically doing another subscribe etc under the hood to get this Ready event.